### PR TITLE
Migrate reCAPTCHA support to Google reCAPTCHA Enterprise (v2 and v3)

### LINKED ISSUE #16138: Support for reCAPTCHA Enterprise at user registration
### Description

This feature would add support for reCAPTCHA Enterprise at user registration.

### Discussion

_No response_

### Motivation

Current implementation is using outdated reCAPTCHA which Google no longer supports.

### Details

_No response_

### PR DESCRIPTION
**Pull request by derlin**
The RegistrationRecaptcha logic is old and targets the first version of Google reCAPTCHA, which is no longer supported.

This commit updates the logic to align with the new **Google reCAPTCHA Enterprise**, with support for both v2 and v3 captchas:

- v2: checkbox-based, visible captcha that is displayed on the page
- v3: score-based, invisible captcha that is hidden on the page, and based on the user's behavior and identity on other sites.

The old configuration (site key, secret key) is replaced with project ID, site key and API key.

Two new optional configurations are:

- a "v3" toggle: there is no easy way to deduce from a site key which  version is enabled. The user hence has to provide this information  to the widget.
- a "min threshold": this is the minimal score a captcha must return  for the registration to proceed. It should be between 0.0 and 1.0,  and is set to the sensible default of 0.7 - "medium high".

The `register.ftl` template is updated as well. Note that in case of invisible captcha, a call to the captcha API must be made to retrieve the captcha token before submitting the form.
This and the new theme attributes will require custom theme maintainer to udpate their code.

The documentation is also updated to reflect the changes.

*About the new reCAPTCHA*

The new reCAPTCHA version works in multiple stages:

1. on the frontend, a JSON script is loaded on the page, and:
   * with v2, an input (populated automatically) is added to the form, with the checkbox,
   * with v3, nothing is shown, so it is up to us to request a captcha   token before submitting the form. This can be done with a button and a callback.
2. on the backend, an assessment must be created and sent to google, along with the captcha token (this is where the API key is necessary),
3. the google assessment response must then be interpreted, to decide if the reCAPTCHA is valid. This entails checking the action name, the "valid" attribute and the score.

Closes: https://github.com/keycloak/keycloak/issues/16138

### CODE REVIEW

File: themes/src/main/resources/theme/base/login/register.ftl
Comment (jonkoops) [inline themes/src/main/resources/theme/base/login/register.ftl:158]: Why has this script been removed?

**Review by jonkoops — CHANGES_REQUESTED** [review]

File: themes/src/main/resources/theme/base/login/register.ftl
Comment (derlin) [reply themes/src/main/resources/theme/base/login/register.ftl:158]: Definitely a mistake, good catch!

File: docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc
Comment (jonkoops) [inline docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc:1]: Thanks for taking the effort to update the docs.

**Comment by jonkoops** [comment]
@andymunro since this PR updates the documentation, can I ask you to give that a pass?

File: themes/src/main/resources/theme/base/login/register.ftl
Comment (jonkoops) [inline themes/src/main/resources/theme/base/login/register.ftl:115]: Looks like this `div` was removed, but should it have been?

File: themes/src/main/resources/theme/base/login/register.ftl
Comment (jonkoops) [inline themes/src/main/resources/theme/base/login/register.ftl:120]: Same for this password reveal button, I don't think this should have been removed.

File: themes/src/main/resources/theme/base/login/register.ftl
Comment (jonkoops) [inline themes/src/main/resources/theme/base/login/register.ftl:132]: Since this condition does not have another branch, I think we can roll this into the `recaptchaRequired` check, so there is only a single `if` statement checking both conditions.

File: themes/src/main/resources/theme/base/login/register.ftl
Comment (jonkoops) [inline themes/src/main/resources/theme/base/login/register.ftl:149]: For new scripts, we default to the [`module`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#applying_the_module_to_your_html) type to get some more [sane defaults](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#other_differences_between_modules_and_standard_scripts). Can we also format this so the JS is on multiple lines and indented?

**Review by jonkoops — CHANGES_REQUESTED** [review]

**Comment by derlin** [comment]
@jonkoops Thank you for the review! I rebased and updated register.ftl as requested. Let me know!

**Review by ssilvert — DISMISSED** [review]
LGTM

File: services/src/main/java/org/keycloak/authentication/forms/RegistrationRecaptcha.java
Comment (pedroigor) [inline services/src/main/java/org/keycloak/authentication/forms/RegistrationRecaptcha.java:253]: Could you elaborate on why we are dropping support for using `recaptcha.net` as an alternative when there are constraints on using the `google.com` domain?

**Review by pedroigor — COMMENTED** [review]
I would say we need something in release notes and migration guide as the changes here might break existing deployments.

@derlin I'm also curious about what you mean about the support level of reCAPTCHA. Do you mean this https://cloud.google.com/recaptcha-enterprise/docs/compare-versions? Or Google is dropping it completely and recommending using v2/v3?

**Review by jonkoops — CHANGES_REQUESTED** [review]

**Comment by ssilvert** [comment]
@derlin There are a couple of failing tests.  It looks like a couple of tests were not updated to reflect help text changes.  I think if you fix those then this will be ready to merge.

See https://github.com/keycloak/keycloak/actions/runs/6854683386/job/18638409570?pr=24223

**Comment by jonkoops** [comment]
@derlin I believe we currently do no not have tests for this functionality. Would this be something you'd be interested in implementing this as well? Doesn't have to be part of this PR, we can handle it separately.

File: themes/src/main/resources/theme/base/login/register.ftl
Comment (derlin) [reply themes/src/main/resources/theme/base/login/register.ftl:154]: done!

**Review by jonkoops — DISMISSED** [review]
LGTM

**Comment by jonkoops** [comment]
@derlin could you squash the commits on your branch and sign off on the final commit (see [DTO check](https://github.com/keycloak/keycloak/pull/24223/checks?check_run_id=18699382265))?

File: services/src/main/java/org/keycloak/authentication/forms/RegistrationRecaptcha.java
Comment (derlin) [reply services/src/main/java/org/keycloak/authentication/forms/RegistrationRecaptcha.java:253]: Good catch! This was a misunderstanding on my part, I added this option back and updated the doc accordingly. Sorry for the mishap.

**Comment by derlin** [comment]
@jonkoops Conerning:
> @derlin I'm also curious about what you mean about the support level of reCAPTCHA. Do you mean this https://cloud.google.com/recaptcha-enterprise/docs/compare-versions? Or Google is dropping it completely and recommending using v2/v3?

To be honest, I always have a hard time parsing Google documentation, as it is so unclear. At first, I understood they only support enterprise for now on, but now looking again it seems I misunderstood. What I implemented is for Enterprise only, but there is still a simpler "v3" option available. The biggest changes are the script added to the HTML (`../enterprise.js` or `/api.js`), and the whole assessment thing with minimum scores are specific to enterprise... (the normal one is simpler).

(I feel so stupid right now).

So basically, I see multiple options:

1. Only support enterprise, which has the same *Free up to 1 million assessments per month* free tier (the only difference is you need to enable billing) - this PR,
2. Support both, and implement all into one class with a toggle (this may make the code quite complex)
3. Support both, in different classes. The non-enterprise v3 could still be the default (and needs to be updated), the Enterprise one in this PR can be used instead if the user choses to.

What do you think?

**Comment by derlin** [comment]
@jonkoops any news on this?

**Comment by jonkoops** [comment]
Sorry, I haven't had time to get back into a review. My backlog has been piling up a lot and we've been focusing on bug fixes. I'll see if I can make some time.

**Review by jonkoops — DISMISSED** [review]
LGTM

**Comment by derlin** [comment]
What are the next steps? Should I rewrite everything to support both recaptcha and recaptcha enterprise? Should I squash all my commits and add a proper DCO? Some guidance on how to get this PR moving would be nice.

**Comment by pedroigor** [comment]
@derlin Yeah, please sign-off the commit and squash so we can run tests and run a 2nd round of review. 

IIRC, one of the key concerns was related to still supporting recaptcha. But I think you have fixed this?

**Comment by derlin** [comment]
@pedroigor well, actually, this PR still **replaces** recaptcha with recaptcha enterprise. I am working on something to allow to support both, but this will require lots of changes (mostly in the documentation). What I propose is basically to have these settings:

Screenshot 2024-01-05 at 19 46 45

The project ID would be the switch between enterprise and regular. The min. score threshold is the only thing that is not shared (action is actually also used in normal (sorry for the error in the screenshot). The second field would be used for either the site secret or the google api key, depending on whether project id is set. Do you think this is ok from the user perspective? (there will be help texts for every option).

Should I rebase and squash/recreate all commits with this version or still add commits on top so reviewers can follow what is happening?

**Comment by pedroigor** [comment]
@derlin Now I'm also confused. I remember testing your PR and having it working with recaptcha. At that time, IIRC my only concern was related to the `recaptcha.net` ...

Perhaps we should just have your initial proposal ready (both versions from the same action) and then evaluate. We will probably end up with distinct actions, I would say. But for now, let's make it work and then evaluate.

Yeah, multiple commits are fine for now.

Does it make sense for you?

**Comment by derlin** [comment]
@pedroigor I pushed the proposition for supporting normal/enterprise and v2/v3. I havent updated the docs for now, waiting for a feedback. I don't know how you tested... What did you put as project ID and api keys? Normal recaptcha only provide two values (site key, secret key), while enterprise require three values (project id, site key and api key).

File: docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc
Comment (andymunro) [inline docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc:20]:
```suggestion
. Create a key for a "Website" platform, and choose the desired key type. Leave the defaults for v3 reCAPTCHA (invisible), or enable *Use checkbox challenge* for a v2 reCAPTCHA (visible). Note the site key for future use in this procedure.
```

**Review by andymunro — CHANGES_REQUESTED** [review]
Looks good, @derlin. Just one change to suggest.

**Comment by pedroigor** [comment]
@derlin Thanks, looking at it this week.

**Comment by derlin** [comment]
> Looks good, @derlin. Just one change to suggest.

I still have to redo the documentation as I now incorporated normal and enterprise support. Waiting for @pedroigor to validate the new code before updating it (we need to clearly explain the new flags and the two "modes").

**Comment by pedroigor** [comment]
Hey @derlin. I did not forget about it :)

It works fine and I can use both recaptcha and recaptcha enterprise without any issue.

However, the UX is far from ideal and quite confusing. Mainly in regards to how enterprise is enabled (based on the project id). The same goes for the score, it is easy to get confused and mess things up until you find the right fields and values you need to set.

That is my only concern.

I've updated your PR with a commit that introduces a separate provider for enterprise. Please, let me know if it works for you.

Sorry for losing your previous commit message and you not being added as the author. Please, feel free to change the commit to have you as the author as well as improve the message if you want to. I had to do that to be able to run tests.

**Comment by pedroigor** [comment]
Hey @derlin, wdyt about the latest changes? Do they work for you?

**Review by jonkoops — DISMISSED** [review]

**Comment by derlin** [comment]
@pedroigor So sorry, these past few weeks have been hectic. I am working on it! I plan to rebase, edit some of your changes, squash and add a commit on top with the fixes to the documentation. Not sure I will be able to do all that before the weekend though. Sounds good?

**Comment by pedroigor** [comment]
@derlin Not a problem at all. Sure, take your time. We have time until we can merge changes for Keycloak 24 (2 weeks).

**Comment by derlin** [comment]
@jonkoops  @andymunro sorry to ping you guys. I wanted to test my changes on the documentation, but impossible to build it. I always get the same `IllegalArgumentException` from the header-maven-plugin. Do you have any clue on how to fix it? I would like to avoid committing something I haven't reviewed properly.

```
mvn install -pl docs/documentation/header-maven-plugin -Pdocumentation
Failed to execute goal org.apache.maven.plugins:maven-plugin-plugin:3.4:helpmojo (header-goal) on project header-maven-plugin: Execution header-goal of goal org.apache.maven.plugins:maven-plugin-plugin:3.4:helpmojo failed.: IllegalArgumentException
```

**Comment by derlin** [comment]
I rebased and cleaned up the documentation, adding a section about recaptcha enterprise for extra clarity. Let me know if this works properly on your end, but I believe we are nearly there 😊 .

**Review by pedroigor — DISMISSED** [review]
If @andymunro is OK with the docs, LGTM.

File: themes/src/main/resources/theme/base/login/register.ftl
Comment (jonkoops) [inline themes/src/main/resources/theme/base/login/register.ftl:94]: A button should always have an explicit `type` attribute, this is to prevent accidental form submission, since the default type is `submit`.

File: themes/src/main/resources/theme/base/login/register.ftl
Comment (jonkoops) [inline themes/src/main/resources/theme/base/login/register.ftl:101]: Let's make things consistent and use a `button` element here as well.

**Review by jonkoops — CHANGES_REQUESTED** [review]
Just a couple of small nits

File: themes/src/main/resources/theme/base/login/register.ftl
Comment (derlin) [reply themes/src/main/resources/theme/base/login/register.ftl:75]: yes it is :) see https://developers.google.com/recaptcha/docs/display#auto_render

Comment (jonkoops) [reply themes/src/main/resources/theme/base/login/register.ftl:75]: Good to know, thanks!

File: themes/src/main/resources/theme/base/login/register.ftl
Comment (derlin) [reply themes/src/main/resources/theme/base/login/register.ftl:94]: Added `type="submit"` to the button.

**Review by jonkoops — DISMISSED** [review]
LGTM, thanks @derlin!

**Comment by derlin** [comment]
@jonkoops What are the next steps? Can I still do something?

**Review by ssilvert — DISMISSED** [review]
Looks like the wrong UI controls are being used for selecting `recaptcha.net` versus `recaptcha v3` versus `none`.

Only one choice should be allowed, right?  So that would mean we should use something like radio buttons or a dropdown.

File: docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc
Comment (andymunro) [inline docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc:22]:
```suggestion
. Create a reCAPTCHA and choose between Challenge v2 (visible checkbox) or Score-based, v3 (invisible) to get your reCAPTCHA site key and secret. Note them down for future use in this procedure.
```

File: docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc
Comment (andymunro) [inline docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc:65]:
```suggestion
. On your Google Cloud Project, go to *Credentials* and create an API key.
```

File: docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc
Comment (andymunro) [inline docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc:69]:
```suggestion
. Navigate to the {project_name} Admin Console.
```

File: docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc
Comment (andymunro) [inline docs/documentation/server_admin/topics/users/proc-enabling-recaptcha.adoc:88]:
```suggestion
.. (Optional) Customize the *Min. Score Threshold* as you see fit. Set it to the minimum score, between 0.0 and 1.0, that a user should achieve on reCAPTCHA to be allowed to register. See https://cloud.google.com/recaptcha-enterprise/docs/interpret-assessment-website#interpret_scores[interpret scores].
```

**Review by andymunro — CHANGES_REQUESTED** [review]
A few minor changes, @derlin

**Comment by derlin** [comment]
> Looks like the wrong UI controls are being used for selecting `recaptcha.net` versus `recaptcha v3` versus `none`.
> 
> Only one choice should be allowed, right? So that would mean we should use something like radio buttons or a dropdown.

I am not sure I understand. Recaptcha.net and v3 are two different things. recaptcha.net is whether to use google.com or recaptcha.net as the domain to load the script, while v2/v3 is the type of captcha to use. They are both independent settings. And what do you mean by "None"?

**Review by jonkoops — DISMISSED** [review]

**Review by andymunro — APPROVED** [review]
Looks good @derlin. Thanks for the updates.

File: themes/src/main/resources/theme/base/login/register.ftl
Comment (pedroigor) [reply themes/src/main/resources/theme/base/login/register.ftl]: @jonkoops This is now breaking a lot of tests. I would propose to have follow-ups to any improvement we want to do as a follow-up. Otherwise, we are losing a bit of focus here ...

@derlin Sorry for the back-and-forth, again. I just tested reverting this part to how it was before (using an input) and it should be enough to pass tests and finally have this one in once the tests are OK.

File: themes/src/main/resources/theme/base/login/register.ftl
Comment (pedroigor) [reply themes/src/main/resources/theme/base/login/register.ftl:94]: I would also avoid doing that now but as a follow-up.

**Comment by jonkoops** [comment]
Looks like we have some failing tests, @derlin could you investigate the cause?

**Review by pedroigor — APPROVED** [review]
@jonkoops I'm reverting the changes that were breaking tests. If you agree, we can improve things as a follow-up.

Run another round of tests and looks good to me.

@derlin Please, let us know if it is working well for you now so we can finally merge this one ...

**Comment by jonkoops** [comment]
Sure, as long as we log an issue to change that input back to a button I am fine with that 👍

**Review by jonkoops — APPROVED** [review]

**Comment by pedroigor** [comment]
@derlin Thanks a lot for your time on this. Looking forward to more improvements in this area :)

**Comment by alexagc** [comment]
Hi @pedroigor @jonkoops, this new implementation (and the older one) with the new user profile breaks profile validation, by  default the max length of unmanaged attributes its 2048 and sometimes the g-recaptcha-response exceed that number breaking the registration. One temporal solution its add that property to user profile and change the default length to one over 2400 (we dont see any higher number).
Maybe the solution its to ignore that property like the password and password confirm on profile validation (https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/forms/RegistrationUserCreation.java#L251)
If its needed i can make a pull request with the changes.

**Comment by jonkoops** [comment]
@alexagc feel free to shoot us a PR, that is always more than welcome. Also make sure to log an issue describing the problem and link it to your PR. Thanks!

**Pull request merged**
